data/blog/lost-in-the-haystack.md

```diff
@@ -0,0 +1,295 @@
+
+```
+┌─query_duration_ms─┬─read_rows─┬─read_bytes─┬─memory_usage─┬─query────────────────┐
+│                24 │    245760 │    7047494 │     14303428 │ WHERE old = 'Safari' │
+│                14 │    286720 │    4874860 │      8743218 │ WHERE new = 'Safari' │
+└───────────────────┴───────────┴────────────┴──────────────┴──────────────────────┘
+```
+
+A pre-requisite to running this query is that you've run _other_ queries against your test table. But once you've done that you should see that our new, encoded column has significantly better performance than the old column. We read more rows with less memory usage and it takes us significantly less time to complete the query. Neat!
+
+Be careful running this migration. It can be tough depending on the size of your dataset.
+
+Side bar. Dropping nullability has implications for your query. For example, the `any` function will ignore nulls when computing its result. If you have an empty string value in your dataset that was previously marked as null your query may return that value rather than the non-empty string it previously returned. There's a simple solution to this. ClickHouse functions accept an `If` combinator and this combinator can be used to strip the empty values. Like so: `anyIf(column, notEmpty(column))`. The second argument to the function is any expression which returns a boolean result.
+
+**Applying Indexes**
+
+ClickHouse indexes are a little different from indexes in other databases. They don't point to a row. They point to a granule which is a collection of rows. At least one row will contain the indexed value for a given matched granule.
+
+This is important because an index is no guarantee of a fast query. If your value has a 1/8192 (or greater) chance of being present on a row then indexing it will give you nothing. ClickHouse will scan every row in the database to answer your query.
+
+Indexes should only be used for values which are rare. A UUID is a great candidate for an index. Someone's age in years is not.
+
+I'll leave applying indexes as an exercise for the reader, we have a different problem we need to solve. Session Replay has an aggregated data model. Indexes won't work. We need an alternative.
+
+One possibility is a sub-query. Our `replay_id` column is indexed and is not under any aggregate function. We can make scalar comparisons against it in the WHERE clause. A sub-query would then need to return a set of replay-ids matching some indexed condition. For example:
+
+```sql
+WHERE replay_id IN (
+    SELECT replay_id
+    FROM replay_events
+    WHERE error_id = 'my_error_id'
+)
+```
+
+For reasons outside the scope of this post, we did not end up using this approach. We decided on an alternative. In our preflight query, under certain conditions, we can apply WHERE clauses which reduce the aggregation set but do not alter the outcome of the query. For example, consider the question "show me every replay which contains the error_id `x`".
```

Comment:
```suggestion
For reasons outside the scope of this post, we did not end up using this approach. We decided on an alternative. In our preflight query, under certain conditions, we can apply WHERE clauses that reduce the aggregation set but do not alter the outcome of the query. For example, consider the question "show me every replay which contains the error_id `x`".
```

Reply:
It doesn't really matter imo, I think because it's referring to a specific object `WHERE clauses` it should be `that` but happy to let this go.